unit/util/piped_process.cpp

```diff
@@ -0,0 +1,272 @@
+/// \file
+/// \author Diffblue Ltd.
+/// Unit tests for checking the piped process communication mechanism.
+
+#ifdef _WIN32
+// No unit tests yet!
+#else
+
+#  include <testing-utils/use_catch.h>
+#  include <util/optional.h>
+#  include <util/piped_process.h>
+#  include <util/string_utils.h>
+
+TEST_CASE(
+  "Creating a sub process and reading its output.",
+  "[core][util][piped_process]")
+{
+  const std::string to_be_echoed = "The Jabberwocky";
+  // Need to give path to avoid shell built-in invocation
+  std::vector<std::string> commands;
+  commands.push_back("/bin/echo");
+  commands.push_back(to_be_echoed);
+  piped_processt process = piped_processt(commands);
+
+  // This is an indirect way to detect when the pipe has something. This
+  // could (in theory) also return when there is an error, but this unit
+  // test is not doing error handling.
+  process.can_receive(PIPED_PROCESS_INFINITE_TIMEOUT);
+  std::string response = strip_string(process.receive());
+
+  REQUIRE(response == to_be_echoed);
+}
+
+TEST_CASE(
+  "Creating a sub process with a binary that doesn't exist.",
+  "[core][util][piped_process]")
+{
+  const std::string expected_error(
+    "Launching jkfadsjfkljdskalfjksdakjfkjdskjflkjasljdflksdjj failed");
+  // Need to give path to avoid shell built-in invocation
+  std::vector<std::string> commands;
+  commands.push_back("jkfadsjfkljdskalfjksdakjfkjdskjflkjasljdflksdjj");
+  piped_processt process = piped_processt(commands);
+
+  // This is an indirect way to detect when the pipe has something. This
+  // could (in theory) also return when there is an error, but this unit
+  // test is not doing error handling.
+  process.can_receive(PIPED_PROCESS_INFINITE_TIMEOUT);
+  std::string response = process.receive();
+  // This tracks how many times we tried, if for some reason we are stuck
+  // give up eventually for this test.
+  int too_many_tries = 0;
+  // The expected length of the output string is 6 characters, keep
+  // trying up to the retry limit of 5 or the string is long enough
+  while(too_many_tries < 5 && response.length() < 64)
+  {
+    // Wait a short amount of time to try and receive
+    process.can_receive(10);
+    response += process.receive();
+    too_many_tries++;
+  }
+
+  REQUIRE(response.substr(0, 64) == expected_error);
+}
+
+TEST_CASE(
+  "Creating a sub process of z3 and read a response from an echo command.",
+  "[core][util][piped_process]")
+{
+  std::vector<std::string> commands;
+  commands.push_back("z3");
```

Comment:
How do we tell CBMC developers who want to run the unit tests that they need Z3?

Reply:
This (and all other tests referencing `z3`) is now hidden behind a *hide tag* `.` - `[.z3]`.

An ordinary invocation of `bin/unit` without any tags will skip all of these tests. I will modify the CI scripts to invoke the unit test binary with the z3 flag passed as well - this we felt is the best compromise between not requiring z3 for the unit tests for an ordinary user, and testing the functionality we need to test in CI as well as locally.